// @ecoeats/application/src/index.ts
// Ports
export { DepotCommandes } from "./ports/DepotCommandes";
export { DepotRestaurants } from "./ports/DepotRestaurants";
export { DepotPlats } from "./ports/DepotPlats";
export { DepotClients } from "./ports/DepotClients";
export { DepotLivreurs } from "./ports/DepotLivreurs";
export { DepotFactures } from "./ports/DepotFactures";
export { ServiceCartographie } from "./ports/ServiceCartographie";
export { ServicePaiement } from "./ports/ServicePaiement";
export { DepotComptes } from "./ports/DepotComptes";

// Use Cases Auth
export { InscriptionUseCase } from "./use-cases/auth/InscriptionUseCase";
export { ConnexionUseCase } from "./use-cases/auth/ConnexionUseCase";

// Use Cases Client
export { ListerRestaurantsUseCase } from "./use-cases/client/ListerRestaurantsUseCase";
export { VoirMenuRestaurantUseCase } from "./use-cases/client/VoirMenuRestaurantUseCase";
export { AjouterAuPanierUseCase } from "./use-cases/client/AjouterAuPanierUseCase";
export { PasserCommandeUseCase } from "./use-cases/client/PasserCommandeUseCase";
export { PayerCommandeUseCase } from "./use-cases/client/PayerCommandeUseCase";
export { ListerCommandesClientUseCase } from "./use-cases/client/ListerCommandesClientUseCase";

// Use Cases Restaurateur
export { AjouterPlatUseCase } from "./use-cases/restaurant/AjouterPlatUseCase";
export { ModifierPlatUseCase } from "./use-cases/restaurant/ModifierPlatUseCase";
export { SupprimerPlatUseCase } from "./use-cases/restaurant/SupprimerPlatUseCase";
export { AccepterCommandeUseCase } from "./use-cases/restaurant/AccepterCommandeUseCase";
export { RefuserCommandeUseCase } from "./use-cases/restaurant/RefuserCommandeUseCase";
export { MarquerCommandePreteUseCase } from "./use-cases/restaurant/MarquerCommandePreteUseCase";
export { ListerCommandesRestaurantUseCase } from "./use-cases/restaurant/ListerCommandesRestaurantUseCase";
export { ModifierRestaurantUseCase } from "./use-cases/restaurant/ModifierRestaurantUseCase";
export { ObtenirMonRestaurantUseCase } from "./use-cases/restaurant/ObtenirMonRestaurantUseCase";

// Use Cases Livreur
export { ChangerStatutLivreurUseCase } from "./use-cases/livreur/ChangerStatutLivreurUseCase";
export { ProposerLivraisonUseCase } from "./use-cases/livreur/ProposerLivraisonUseCase";
export { AccepterLivraisonUseCase } from "./use-cases/livreur/AccepterLivraisonUseCase";
export { RefuserLivraisonUseCase } from "./use-cases/livreur/RefuserLivraisonUseCase";
export { ObtenirPropositionsLivreurUseCase } from "./use-cases/livreur/ObtenirPropositionsLivreurUseCase";
export { AttribuerLivraisonUseCase } from "./use-cases/livreur/AttribuerLivraisonUseCase";
export { TerminerLivraisonUseCase } from "./use-cases/livreur/TerminerLivraisonUseCase";
export { ObtenirLivreurUseCase } from "./use-cases/livreur/ObtenirLivreurUseCase";
